### DISCUSSION TITLE
/combining-zfs-with-other-filesystems-using-mergerfs/

### DISCUSSION THREAD

**Discussion by migratecomments** [ironicbadger/blog.ktz.me#18] [Announcements]
Combining ZFS with other filesystems using mergerfs
        
[https://blog.ktz.me/combining-zfs-with-other-filesystems-using-mergerfs/](https://blog.ktz.me/combining-zfs-with-other-filesystems-using-mergerfs/)

**Comment by migratecomments** [comment]
```header
Originally posted by donnib on 2021-02-16 12:25
```

Hi Alex,  
So here are some questions in general for a home server i am building. I have at the moment 3,5tb disk space on a Synology and i'd like to move away from that to a new server i just got, an older HP DL360P G8 with 8 bays of 2.5" disks. I'd like to make a ZSH pool for my precious data then use mergerfs just as you describe in this post BUT here are some more questions.

First i want to state that i'd like to use Proxmox and i have a need to run a Plex server, a bunch of docker containers and be able to share data locally aka NAS so my questions are :

1\. Where would you install proxmox ? On a SSD and use one of the bays ?  
2\. Where would you store the images for the VM ?  
3\. Where would the mergerfs/Snapraid be installed ? on the proxmox host ?  
4\. Where would the samba server be ? on the proxmox host ?

\- I think 2tb for my important data is enough so i am guessing two 2.5" disks of 2tb would be ok in ZFS, what do you think ?  
\- One bay used for one system disk (SSD) running proxmox  
\- then i have 5 bays left for other data that i can use with mergerfs and snapraid, i don't have a need to use all bays but few disks are probably ok for media.

I am sure i have more questions as soon i hear from you ;)

Thx

**Reply by migratecomments** [reply]
```header
Originally posted by Josef Karlsson on 2021-03-27 04:15
```

I'm far from an expert in this area, but I'm currently setting up a somewhat similar setup.

I like to keep my proxmox host as clean as possible, with minimal additional installs. My solution solution instead is running proxmox on ZFS mirror (2.5 disks) as base system. I then created a LXC container and mounted the mixed sized drives to this.

From the container I then run mergerfs as well as docker and samba. This gives me full control and flexibility, as well as security for the system.

In my use case I don't see the point including zfs with mergerfs. Since the container it self is a mounted storage on the ZFS, all the data in the container (except the storage drives) would be protected.

I use my SSDs VM:s

**Comment by digitalfrost84** [comment]
While this is a really nice idea, the biggest gotcha for me here has to be the fact that you cannot move files in /mnt/storage for example from the nonredundant to the redundant part. See "'mv /mnt/pool/foo /mnt/disk1/foo' removes 'foo'" in mergerfs documentation: https://github.com/trapexit/mergerfs

If you move a file between disks, it will copy the file and then afterwards since it was moved, delete the source. That's how a move works right?! However mergerfs will see if it finds the path and it will find it in the new supposedly redundant destination, and then delete the file. I find this highly risky to use for anything important.

I would like to build something where I have a "new" or "incoming" folder for any data that is nonredundant by default, and then an archive folder where I move things that I want to keep which should be redundant then.

**Reply by matth-av** [reply]
I don't think this is an issue. Say you have `/mnt/redundant` and `/mnt/nonredundant` which are merged into `/mnt/storage`, and you want to move the file you would just do `mv /mnt/redundant/file /mnt/nonredundant/file` and mergerfs isn't involved in the transaction at all. You're probably right you shouldn't go from `/mnt/storage` into `/mnt/nonredundant` but that honestly seems like a very strange thing to do. You shouldn't want to interact with the mapped filesystem in the same process as the raw one. Seems like weird recursion would incur.

**Reply by digitalfrost84** [reply]
I agree one should not do stupid things, the problem is that basic assumptions about how file systems work (moving a file will not delete it) are not true when having mergerfs inbetween. If you always remember this and/or you have backups your data is safe, but one mistake and it will delete files.

**Reply by matth-av** [reply]
I think a user moving files around shouldn't have access to the underlying disks anyway, so only sudoers and scripts are moving files between filesystems. In daily use of files the mergerfs mount is just a pool of storage to use. Like with ZFS you wouldn't be thinking about the vdevs inside your pool in the same time as a user moving a file. I don't know if that made much sense but if you share only the mergerfs mount to users then it shouldn't increase any risk of accidentally deleting files.
